### DISCUSSION TITLE
Automatic detection of an LLVM installation in xrepo

### DISCUSSION THREAD

**Discussion by A2va** [xmake-io/xmake#3194] [Ideas]
### Is your feature request related to a problem? Please describe.

Currently, `xmake f --toolchains=llvm` display `error: llvm toolchain not found!` even if LLVM is installed with xrepo. LLVM binaries need to be on the PATH, xmake cannot detect that xrepo already installed the LLVM package.

### Describe the solution you'd like

Instead, xmake can detect the llvm package from xrepo. The same principle could be applied with clang.

**Comment by waruqi** [comment]
see https://github.com/xmake-io/xmake/blob/master/tests/projects/package/toolchain_llvm/xmake.lua

https://github.com/xmake-io/xmake/blob/master/tests/projects/package/toolchain_llvm_mingw/xmake.lua

**Comment by A2va** [comment]
Thanks, but is it possible that xmake detects the presence of the LLVM package when running `xmake config --toolchains=llvm`

**Comment by waruqi** [comment]
> Thanks, but is it possible that xmake detects the presence of the LLVM package when running `xmake config --toolchains=llvm`

It did.

https://github.com/xmake-io/xmake-repo/blob/a54337a96d3aa2e6b9fbfd5f71aacbd59e02f27a/packages/l/llvm/xmake.lua#L74

**Comment by A2va** [comment]
I installed LLVM with xrepo:
```
PS C:\Users\leres\Documents\xmake-vscode-test\cpp> xrepo scan llvm
scanning packages ..
llvm-14.0.0:
  -> dd3b83b656aa4deda4407126fd76a067: windows, x64, unused
    -> {all=false,bolt=false,clang=true,["clang-tools-extra"]=false,["compiler-rt"]=true,debug=false,flang=false,libc=false,libclc=false,libcxx=true,libcxxabi=true,libunwind=true,lld=false,lldb=false,mlir=false,openmp=false,polly=false,pstl=false,shared=false}
```
But why xmake doesn't detect it ?

```
PS C:\Users\leres\Documents\xmake-vscode-test\cpp> xmake f --toolchain=clang
checking for platform ... windows
checking for architecture ... x64
error: target(cpp): toolchain not found!

PS C:\Users\leres\Documents\xmake-vscode-test\cpp> xmake f --toolchain=llvm
checking for platform ... windows
checking for architecture ... x64
error: llvm toolchain not found!
```

**Reply by waruqi** [reply]
see https://github.com/xmake-io/xmake/blob/master/tests/projects/package/toolchain_llvm/xmake.lua 

use add_requires or set `xmake f --sdk=`

**Reply by A2va** [reply]
True, but in the cmd line, setting the llvm  path with `xmake f --sdk=` is not very convenient. Knowing that xrepo already has its location.

**Reply by waruqi** [reply]
so you can use add_requires, it will find it.

**Reply by A2va** [reply]
So there's no way xmake can find llvm from the command line? Only by adding `add_requires` to the xmake.lua?

**Reply by waruqi** [reply]
> So there's no way xmake can find llvm from the command line? Only by adding `add_requires` to the xmake.lua?

why not use add_requires, it will find system llvm first. 

> So there's no way xmake can find llvm from the command line?

--sdk or you add llvm to $PATH

**Reply by A2va** [reply]
| why not use add_requires, it will find system llvm first.
It might be convenient to change the toolchain without having to add it to the PATH or modify the xmake.lua

Thank you very much for your help, I understand better now.
